torch/nn/modules/module.py

```diff
@@ -193,13 +193,36 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        def compute_should_use_data(tensor, tensor_applied):
+            if tensor.has_same_tensorimpl_type_as(tensor_applied):
+                # If the new tensor has the same TensorImpl type as the existing tensor,
+                # then we take `torch.__future__.overwrite_module_params_on_conversion()`
+                # into account.
+                return not torch.__future__.overwrite_module_params_on_conversion()
+            else:
+                # If the new tensor has a different TensorImpl type, we always overwrite
+                # the existing tensor.
+                return False
+
+        for key, param in self._parameters.items():
             if param is not None:
-                # Tensors stored in modules are graph leaves, and we don't
-                # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                if compute_should_use_data(param, fn(param)):
```

Comment:
I don't think executing `fn` twice is a good idea, this is really going to slow down moving large models and it's bogus for inplace functions.